File: py/kintsugi/checks.py
"""Running check commands and turning their output into findings.

Checks run with cwd = source root, shell-resolved, so `pytest` and `ruff`
find the repo's own toolchain. A non-zero exit with no parseable finding is
a *crash*, not a defect: the loop reports it and refuses to heal it, because
a repair loop that "heals" a broken harness is a repair loop that rewrites
working code because its own plumbing failed.
"""

import os
import re
import subprocess
import time

from .parsers import parse_lines, parse_radon, parse_rust, parse_strict

# ANSI CSI color sequences. CI hosts force tool color (rustup actions set
# CARGO_TERM_COLOR=always), and colored diagnostics carry escape codes before
# the `warning:`/`-->`/`panicked at` anchors every parser relies on:
#
#   \x1b[1m\x1b[93mwarning\x1b[0m: unused import: `std::fmt`
#    \x1b[1m\x1b[96m--> \x1b[0msrc/lib.rs:1:5
#
# Output is sanitized at this one funnel so every parser sees plain text.
_ANSI_RE = re.compile(r"\x1b\[[0-9;?]*[ -/]*[@-~]")


def _strip_ansi(s: str) -> str:
    return _ANSI_RE.sub("", s)


def _parse(defn, output: str, cwd: str):
    parser = defn.get("parser")
    if parser == "strict":
        return parse_strict(output, cwd, defn["name"])
    if parser == "lines":
        return parse_lines(output, cwd, defn["name"])
    if parser == "rust":
        return parse_rust(output, cwd, defn["name"])
    if parser == "radon":
        return parse_radon(output, cwd, defn["name"])
    # tsc/tap/spec output shapes belong to the Node engine; a config that
    # declares them here is a config that cannot be read.
    return []


def run_check(defn, cwd: str, timeout_ms: int | None = None):
    """Run one check command and return a CheckResult-shaped dict."""
    if timeout_ms is None:
        # A check can declare a longer timeout for cold-building
        # toolchains (cargo compiles the crate before testing).
        timeout_ms = defn.get("timeoutMs", 120_000)
    started = time.time()
    env = dict(os.environ)
    env.pop("NODE_TEST_CONTEXT", None)

    try:
        proc = subprocess.Popen(
            defn["command"],
            shell=True,
            cwd=cwd,
            env=env,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
        )
    except OSError as err:
        return {
            "check": defn["name"],
            "findings": [],
            "exitCode": -1,
            "durationMs": int((time.time() - started) * 1000),
            "crashed": True,
            "output": f"failed to start: {err}",
        }

    killed = False
    try:
        raw, _ = proc.communicate(timeout=timeout_ms / 1000)
        code = proc.returncode
    except subprocess.TimeoutExpired:
        proc.kill()
        raw, _ = proc.communicate()
        code = -2
        killed = True

    # Decode UTF-8 explicitly: the engines' own scanners (and tools like
    # ruff) emit UTF-8, but text=True would decode with the *locale* (cp1252
    # on Windows) and mangle every non-ASCII byte — corrupting messages and
    # with them the fingerprints the ledger keys on. This mirrors the TS
    # engine's String(buffer) decode, so both engines agree byte-for-byte.
    output = _strip_ansi((raw or b"").decode("utf-8", errors="replace"))
    duration_ms = int((time.time() - started) * 1000)
    # Exit 0 means the check passed — its output may still *look* like
    # findings, but a passing check contributes no findings by definition.
    # Except when the check declares parseOnExit0 (analysis tools like
    # radon always exit 0 — their findings live in the text, not the
    # status). pytest exits 5 with "no tests ran" for an empty suite — that
    # is a clean state, not a defect and not a broken harness, and it must
    # not block the verify gate (test generation exists precisely to give a
    # testless repo its first tests).
    empty_suite = re.search(r"no tests ran", output, re.IGNORECASE) is not None
    parsed = [] if empty_suite or (code == 0 and not defn.get("parseOnExit0")) else _parse(defn, output, cwd)

    # A check owns its defect class: filterCodes keeps only the codes it was
    # configured for, and a configured severity overrides the parser default.
    findings = parsed
    if defn.get("filterCodes"):
        allowed = set(defn["filterCodes"])
        findings = [f for f in findings if f.get("code") in allowed]
    if defn.get("severity"):
        findings = [{**f, "severity": defn["severity"]} for f in findings]

    return {
        "check": defn["name"],
        "findings": findings,
        "exitCode": code,
        "durationMs": duration_ms,
        "crashed": killed or (not empty_suite and code != 0 and len(parsed) == 0),
        "output": output,
    }

File: src/checks.ts
import { spawn } from 'node:child_process';
import type { CheckDef, CheckResult } from './types.js';
import { parseTsc, parseTap, parseLines, parseSpec } from './parsers.js';

/**
 * Run one check command and turn its output into findings.
 *
 * Checks run with cwd = source root, shell-resolved, so `npm run typecheck`
 * finds the repo's own toolchain. A non-zero exit with no parseable finding
 * is a *crash*, not a defect: the loop reports it and refuses to heal it,
 * because a repair loop that "heals" a broken harness is a repair loop that
 * rewrites working code because its own plumbing failed.
 */
export function runCheck(
  def: CheckDef,
  cwd: string,
  timeoutMs = 120_000,
): Promise<CheckResult> {
  return new Promise((resolvePromise) => {
    const started = Date.now();
    let output = '';
    let killed = false;

    // Checks are external commands and must behave as if a scheduler ran
    // them — never as a nested child of whatever harness Kintsugi itself is
    // running under. Without this, a `node --test` parent sets
    // NODE_TEST_CONTEXT and the repo's own `npm test` silently skips its
    // files ("recursively within a test file"), exiting 0 with no results.
    const env = { ...process.env };
    delete env.NODE_TEST_CONTEXT;
    const child = spawn(def.command, { cwd, shell: true, windowsHide: true, env });
    const timer = setTimeout(() => {
      killed = true;
      child.kill();
    }, timeoutMs);

    child.stdout.on('data', (d) => { output += d; });
    child.stderr.on('data', (d) => { output += d; });

    child.on('error', (err) => {
      clearTimeout(timer);
      resolvePromise({
        check: def.name,
        findings: [],
        exitCode: -1,
        durationMs: Date.now() - started,
        crashed: true,
        output: `failed to start: ${err.message}`,
      });
    });

    child.on('close', (code) => {
      clearTimeout(timer);
      const durationMs = Date.now() - started;
      // Exit 0 means the check passed — its output may still *look* like
      // findings (a version check that prints "version: ok"), but a passing
      // check contributes no findings by definition.
      // Crash means *no typed output at all* — a broken harness. Exit non-zero
      // with parseable findings is a defect, even when the filter below drops
      // every one of them (another check owns that defect class).
      const parsed = code === 0 ? [] : parseWithFallback(def, output, cwd);
      // A check owns its defect class: tsc-based lint only keeps the codes
      // it was configured for, so the same type error is not reported twice
      // by two checks and repaired (or quarantined) twice.
      const findings = def.filterCodes?.length
        ? parsed.filter((f) => def.filterCodes!.includes(f.code ?? ''))
        : parsed;
      resolvePromise({
        check: def.name,
        findings,
        exitCode: killed ? -2 : (code ?? -1),
        durationMs,
        crashed: killed || (code !== 0 && parsed.length === 0),
        output,
      });
    });
  });
}

function parse(def: CheckDef, output: string, cwd: string) {
  switch (def.parser) {
    case 'tsc': return parseTsc(output, cwd, def.name);
    case 'tap': return parseTap(output, cwd, def.name);
    case 'lines': return parseLines(output, cwd, def.name);
    default: return [];
  }
}

/**
 * A check that exits non-zero must still speak in typed findings. A repo
 * whose `npm test` runs `tsx --test` without a reporter emits *spec*
 * output (✔/✖) when piped on newer Node — indistinguishable from a crash
 * to a TAP-only parser. If the declared parser yields nothing but the
 * output is clearly spec-shaped, parse it as spec: a failing test is a
 * defect, never a broken harness.
 */
function parseWithFallback(def: CheckDef, output: string, cwd: string) {
  const parsed = parse(def, output, cwd);
  if (parsed.length === 0 && /^[✔✖ℹ]/m.test(output)) {
    return parseSpec(output, cwd, def.name);
  }
  return parsed;
}
